smart_card_connector_app/src/application_unittest.cc

```diff
@@ -721,6 +748,90 @@ TEST_F(SmartCardConnectorApplicationSingleClientTest,
+  // Arrange:
+  StartApplication();
+  SetUpJsClient();
+  SetUpSCardContext();
+
+  // Act:
+  // Simulate connecting a reader.
+  TestingSmartCardSimulation::Device device;
+  device.id = 123;
+  device.type = TestingSmartCardSimulation::DeviceType::kGemaltoPcTwinReader;
+  SetUsbDevices({device});
+  // Request SCardGetStatusChange to check it observes the change.
+  std::vector<Value> reader_states;
+  EXPECT_EQ(SimulateGetStatusChangeCallFromJsClient(
+                kFakeHandlerId, scard_context(),
+                /*timeout=*/INFINITE,
+                ArrayValueBuilder()
+                    .Add(DictValueBuilder()
+                             .Add("reader_name", kPnpNotification)
+                             .Add("current_state", SCARD_STATE_UNAWARE)
+                             .Get())
+                    .Get(),
+                reader_states),
+            SCARD_S_SUCCESS);
+
+  // Assert:
+  ASSERT_THAT(reader_states, SizeIs(1));
+  EXPECT_THAT(reader_states[0], DictSizeIs(4));
+  EXPECT_THAT(reader_states[0], DictContains("reader_name", kPnpNotification));
+  EXPECT_THAT(reader_states[0],
+              DictContains("current_state", SCARD_STATE_UNAWARE));
+  EXPECT_THAT(reader_states[0],
+              DictContains("event_state", SCARD_STATE_CHANGED));
+  EXPECT_THAT(reader_states[0], DictContains("atr", Value::Type::kBinary));
+}
+
+// Test `SCardGetStatusChange()` call from JS detects when a reader is
+// unplugged.
+TEST_F(SmartCardConnectorApplicationSingleClientTest,
+       SCardGetStatusChangeDeviceRemoving) {
+  // Arrange: start with a single device.
+  TestingSmartCardSimulation::Device device;
+  device.id = 123;
+  device.type = TestingSmartCardSimulation::DeviceType::kGemaltoPcTwinReader;
+  SetUsbDevices({device});
+  StartApplication();
+  SetUpJsClient();
+  SetUpSCardContext();
+
+  // Act:
+  // Simulate disconnecting a reader.
+  SetUsbDevices({});
+  // Request SCardGetStatusChange to check it observes the change.
+  std::vector<Value> reader_states;
+  EXPECT_EQ(SimulateGetStatusChangeCallFromJsClient(
+                kFakeHandlerId, scard_context(),
+                /*timeout=*/INFINITE,
+                ArrayValueBuilder()
+                    .Add(DictValueBuilder()
+                             .Add("reader_name", "Gemalto PC Twin Reader 00 00")
+                             .Add("current_state", SCARD_STATE_EMPTY)
+                             .Get())
+                    .Get(),
+                reader_states),
+            SCARD_S_SUCCESS);
+
+  // Assert:
+  ASSERT_THAT(reader_states, SizeIs(1));
+  EXPECT_THAT(reader_states[0], DictSizeIs(4));
+  EXPECT_THAT(reader_states[0],
+              DictContains("reader_name", "Gemalto PC Twin Reader 00 00"));
+  EXPECT_THAT(reader_states[0],
+              DictContains("current_state", SCARD_STATE_EMPTY));
+  EXPECT_THAT(
+      reader_states[0],
+      DictContains("event_state", SCARD_STATE_CHANGED | SCARD_STATE_UNKNOWN |
```

Comment:
Hmm apparently this test is flaky - the `SCARD_STATE_UNKNOWN` flag is sometimes not returned by the call. I actually think this flag should never be returned in this combination - might be a bug in PC/SC-Lite...